intf/intf.go

```diff
@@ -0,0 +1,156 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package intf implements mechanisms to access underlying interfaces in
+// a manner that is agnostic to the underlying implementation. An implementation
+// is included that uses netlink to write to a Linux kernl implementation.
+// Additional implementations can be provided to allow writing to e.g. a
+// gRPCwire implementation.
+package intf
+
+import (
+	"fmt"
+	"net"
+)
+
+// Interface represents information corresponding to a single interface on the
+// underlying system.
+type Interface struct {
+	// Index is an integer index used to refer to the interface.
+	Index int
+	// Name is the name used to refer to the interface in the system.
+	Name string
+	// MAC is the MAC address of the interface.
+	MAC net.HardwareAddr
+}
+
+// ARPEntry is a representation of an ARP entry on the underlying system.
+type ARPEntry struct {
+	// IP is the IP address of the neighbour.
+	IP net.IP
+	// Interface is the name of the interface on which the ARP entry was learnt.
+	Interface string
+	// MAC is the MAC address of the neighbour.
+	MAC net.HardwareAddr
+}
+
+// ARPEvent is used to describe a particular event on the ARP table.
+type ARPEvent int64
+
+const (
+	// ARPUnknown is used when the event is not an add or delete.
+	ARPUnknown ARPEvent = iota
+	// ARPAdd indicates that the ARP neighbour was added.
+	ARPAdd
+	// ARPDelete inidcates that the ARP neighbour was deleted.
+	ARPDelete
+)
+
+// ARPUpdate is used to describe a change to the ARP table.
+type ARPUpdate struct {
+	// Type indicates whether the event was an add or delete.
+	Type ARPEvent
+	// Neigh is the neighbour that the change corresponds to
+	Neigh ARPEntry
+}
+
+// NetworkAccessor is an interface implemented by the underlying system access
+// interface. Through implementing the NetworkAccessor interface it is possible
+// to use the functions within this package whilst using a different underlying
+// implementation (e.g., netlink on Linux to access the kernel).
+type NetworkAccessor interface {
+	// Links returns a set of links that are present on the local system.
+	Interfaces() ([]*Interface, error)
+	// Link retrieves the link with the specified name.
+	Interface(name string) (*Interface, error)
+	// AddressList lists the IP addresses configured on a particular interface.
+	InterfaceAddresses(name string) ([]*net.IPNet, error)
+	// AddressAdd adds address ip to the interface name.
+	AddInterfaceIP(name string, ip *net.IPNet) error
+	// ARPList lists the set of ARP neighbours on the system.
+	ARPList() ([]*ARPEntry, error)
+	// ARPSubscribe writes changes to the ARP table to the channel updates, and uses
+	// done to indicate that the subscription is complete.
+	ARPSubscribe(updates chan ARPUpdate, done chan struct{}) error
+}
+
+// unimplementedAccessor is an accessor interface that returns unimplemented
+// for all underlying methods.
+type unimplementedAccessor struct{}
```

Comment:
note for offline discussion - I want to understand this golang pattern a bit more. why the need for this?